### DISCUSSION TITLE
Dependency resolution metadata. How is it gathered?

### DISCUSSION THREAD

**Discussion by stefanoborini** [python-poetry/poetry#3685] [Q&A]
- [X] I have searched the [issues](https://github.com/python-poetry/poetry/issues) of this repo and believe that this is not a duplicate.
- [X] I have searched the [documentation](https://python-poetry.org/docs/) and believe that my question is not covered.

## Issue

I tried to navigate the code but I could not find it. Maybe you can give me a better answer.

I suspect that the dependency resolver in poetry needs to know the dependency tree, and possibly backtrack if some constraints are not satisfied, but since the dependency tree depends on the version of the package downloaded (a package v1.2 might have different dependencies from v1.3), the tree must be rebuilt at every backtrack, at least for those parts that do not satisfy a found constraint.

How is the dependency list and constraints gathered? do you download the packages, unpack them every time, and get the subdependency information from the setup.py/pyproject.toml? Are they cached, or is an index built somewhere? As far as I know, pypi does not provide metadata files, hence the question.

**Comment by abn** [comment]
1. If package is on PyPI, the package information available at `https://pypi.org/pypi/{package}/json`. https://github.com/python-poetry/poetry/blob/3a789a7a2d9d82d57d388957f70cc985fa19d22f/poetry/repositories/pypi_repository.py#L199-L218
2. If package is in a private repository, the simple endpoint is used. (eg: https://pypi.org/simple/poetry-core/). https://github.com/python-poetry/poetry/blob/3a789a7a2d9d82d57d388957f70cc985fa19d22f/poetry/repositories/legacy_repository.py#L347-L420
3. If metadata is not availably, filename is used to determine python constraints if any. Then the file is downloaded, and inspected. https://github.com/python-poetry/poetry/blob/3a789a7a2d9d82d57d388957f70cc985fa19d22f/poetry/inspection/info.py#L580-L590
4. If it is a wheel, static metadata is extracted.
5. If it is an sdist, it is unpacked and any static metadata is collected (either via poetry pyproject.toml if applicable or sdist metadata if not). If static metadata is unavailable or incomplete, the sdist is metadata is constrcuted via PEP517 metadata build.
6. If the above fails, and project has a setup.py file, we parse it. https://github.com/python-poetry/poetry/blob/3a789a7a2d9d82d57d388957f70cc985fa19d22f/poetry/inspection/info.py#L297-L352

**Reply by stefanoborini** [reply]
wow. Thanks.

**Reply by abn** [reply]
Hope that answers your question. :)
